Add interactive run that combines run and attach.

brendandburns opened the pull request:
@jlowdermilk @bgrant0607 @smarterclayton 

Fixes #10235

brendandburns committed:
Add interactive run that combines run and attach.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -66,10 +70,13 @@ func NewCmdRun(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Int("hostport", -1, "The host port mapping for the container port. To demonstrate a single-machine container.")
 	cmd.Flags().StringP("labels", "l", "", "Labels to apply to the pod(s).")
+	cmd.Flags().BoolP("interactive", "i", false, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
+	cmd.Flags().Bool("tty", false, "Allocated a TTY for each container in the pod.")
```
exec and attach using "t" as the short alias. We should do that everywhere or nowhere. 

Does run support `-o template`?

brendandburns replied:
yes, that's the problem.  `run` includes the printer flags, and the printer flags have `-t` as a short hand for `--template`

We should deprecate `-t` for `--template` but that is a breaking change (as is removing `-t` from `exec`)

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -66,10 +70,13 @@ func NewCmdRun(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().StringP("labels", "l", "", "Labels to apply to the pod(s).")
+	cmd.Flags().BoolP("interactive", "i", false, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
+	cmd.Flags().Bool("tty", false, "Allocated a TTY for each container in the pod.")
+	cmd.Flags().Bool("attach", false, "If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.  Default false, unless '-i/--interactive' is set, in which case the default is true.")
```
This makes sense to me also,  but we should document how we differ from Docker's -d/-a in docs/user-guide/docker-cli-to-kubectl.md, and link to https://docs.docker.com/reference/run/

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -66,10 +70,13 @@ func NewCmdRun(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Int("port", -1, "The port that this container exposes.")
 	cmd.Flags().Int("hostport", -1, "The host port mapping for the container port. To demonstrate a single-machine container.")
 	cmd.Flags().StringP("labels", "l", "", "Labels to apply to the pod(s).")
+	cmd.Flags().BoolP("interactive", "i", false, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
```
In exec and attach, "i" was a short alias for "stdin". We should be consistent and either change this flag to "stdin" or change the others to "interactive".

brendandburns replied:
switched to stdin

derekwaynecarr commented:
If I specify that I want to attach as part of run, should I be restricted to not setting a --replicas flag?

Or is the idea that you just get the first pod, and since this is backed by an rc and you close the connection, the container will just restart for you?

bgrant0607 commented:
Please cite relevant issues when sending PRs. This fixes #10235.

Restricting replicas: good idea.

There is also the question of what to do if the container restarts.

Related -- support for --restart=Never and --restart=OnFailure pods: #4366. We could start with raw pods and add support for Job once it exists. Raw pods would be restricted to 1 replica.

bgrant0607 commented:
Please try to be consistent across kubectl and API changes. Lack of consistency causes significant user and developer pain. You've previously made this argument yourself -- consistency makes it easier for a user to apply what they learn about one part of the system to other parts.

brendandburns commented:
Switched to `--stdin` added a check for `--replicas != 1`, we can't do `-t` for the reasons I explained (unless we're willing to take a breaking change)

Added more documentation.

I'm ok with switching this to be a raw pod for interactive, with a restart policy of never.

I was planning on making that an optional thing for users to choose using the flags you suggested as a follow on PR.

Let me know if you want me to include that in this PR.

Thanks!
--brendan

brendandburns commented:
(and regarding container restarts, I think that is what `kubectl attach ...` is for.  I don't want to stay permanently attached to different containers across restarts (for now at least)

bgrant0607 commented:
We can't eliminate it, but we should deprecate `-t` for the template and remove it from examples and scripts. That can be a separate PR. Please file an issue if you won't get to it right away.

Restart policy can also be a separate PR, assuming that the PRs don't get split across releases.

bgrant0607 commented on docs/user-guide/docker-cli-to-kubectl.md:
````diff
@@ -80,6 +80,17 @@ nginx-http   run=nginx-app   run=nginx-app             80/TCP
+kubectl run [-i] [--tty] --attach <name> --image=<image>
+```
+
+Unlike `docker run ...`, if `--atach` is specified, we attach to `stdin`, `stdout` and `stderr`, there is no ability to control which streams are attached (`docker -a ...`).
````
Typo: atach

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -66,10 +70,13 @@ func NewCmdRun(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().Int("hostport", -1, "The host port mapping for the container port. To demonstrate a single-machine container.")
 	cmd.Flags().StringP("labels", "l", "", "Labels to apply to the pod(s).")
+	cmd.Flags().BoolP("stdin", "i", false, "Keep stdin open on the container(s) in the pod, even if nothing is attached.")
+	cmd.Flags().Bool("tty", false, "Allocated a TTY for each container in the pod.")
```
Please add a comment about why `-t` isn't an alias for --tty for run.

brendandburns commented:
Comments addressed.

Marked `--template/t` deprecated, there doesn't seem to be a good way to just mark the shortcut deprecated, and we want to switch to golang/jsonpath specific templates anyway.

Added issue #12248 for complete removal and doc changes.

Will do restart/never in a different PR.

bgrant0607 commented:
Thanks. LGTM.

bgrant0607 commented:
Mungedocs failed.

brendandburns commented:
weird, I ran it locally and it's clean.  re-pushed.

bgrant0607 commented:
Now gofmt failed.

bgrant0607 commented:
I think a previous PR broke the build.

bgrant0607 commented:
No, it passed: https://travis-ci.org/GoogleCloudPlatform/kubernetes/builds/74163129

bgrant0607 commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -81,7 +83,10 @@ func OutputVersion(cmd *cobra.Command, defaultVersion string) string {
 func PrinterForCommand(cmd *cobra.Command) (kubectl.ResourcePrinter, bool, error) {
 	outputFormat := GetFlagString(cmd, "output")
-	templateFile := GetFlagString(cmd, "template")
+	templateFile := GetFlagString(cmd, "golang-template")
```
@thockin had an opinion on this.

bgrant0607 replied:
Alternatives:
1. `-o template` and different flags for each template format: `--golang-template=...`, `--jsonpath-template=...`, etc.
2. `-o golang-template`, `-o jsonpath-template`, etc. with one template flag ``--template=...`.
3. `-o golang-template=...`, `-o jsonpath-template=...`, etc.

I think @thockin is in favor of (3). It does seem simpler.

smarterclayton replied:
I tend to 3 as well, although the names are pretty long to type :)

On Aug 6, 2015, at 1:18 AM, Brian Grant <notifications@github.com> wrote:

In pkg/kubectl/cmd/util/printing.go
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12220#discussion_r36381964>
:

> @@ -81,7 +83,10 @@ func OutputVersion(cmd *cobra.Command, defaultVersion string) string {
>  // Requires that printer flags have been added to cmd (see AddPrinterFlags).
>  func PrinterForCommand(cmd *cobra.Command) (kubectl.ResourcePrinter, bool, error) {
>  	outputFormat := GetFlagString(cmd, "output")
> -	templateFile := GetFlagString(cmd, "template")
> +	templateFile := GetFlagString(cmd, "golang-template")

Alternatives:
1. -o template and different flags for each template format:
--golang-template=..., --jsonpath-template=..., etc.
2. -o golang-template, -o jsonpath-template, etc. with one template flag `
--template=....
3. -o golang-template=..., -o jsonpath-template=..., etc.

I think @thockin <https://github.com/thockin> is in favor of (3). It does
seem simpler.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12220/files#r36381964>
.

thockin replied:
We eschew obscure abbreviations, but  I will counter the "it's too long" with:

-o go-tmpl="..."
-o jp-tmpl="..."

or 

-o template=jsonpath:"..."
-o template=go:"..."

not sure those are better.

bgrant0607 replied:
Can shell completion deal with `-o <prefix>`?

`go-template` would be ok, rather than `golang-template`, and `jsonpath` rather than `jsonpath-template`. 

"jp" and "tmpl" are a bit too obscure, except maybe as aliases for the verbose forms.

cc @daizouzhou, since this thread is otherwise hard to find

brendandburns commented:
Rebased and repushed.

@thockin I'd really like to get this in, so if you have an opinion, please let me know asap.

brendandburns commented:
@bgrant0607 @smarterclayton @thockin 

Given that this is degrading into a bikeshed, and implementing any of the suggestions is beyond the scope of this PR, can I remove the `--template` deprecation from this PR and we can address it in a different PR?

brendandburns commented:
PR changed after LGTM, removing LGTM.

brendandburns commented:
@thockin @bgrant0607 
Ok, I removed the deprecation of `--template` since that's devolving and really unrelated to this PR, we can do it in a different PR.

PTAL
--brendan

bgrant0607 commented:
Punting the --template change to another PR SGTM.

bgrant0607 commented:
I still see --golang-template in this PR.

bgrant0607 commented:
gofmt failed

brendandburns commented:
Rebased and re-pushed, all tests are passing.

Thanks
--brendan

bgrant0607 commented:
LGTM
